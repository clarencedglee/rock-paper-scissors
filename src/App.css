body {
  background: #EAE2D7;
  color: #717171;
  font-family: 'Open Sans', sans-serif;
}

h1 {
  font-size: 60px;
  letter-spacing: -0.05em;
  line-height: 0.8;
  margin: 0;
  margin-left: -0.06em;
  text-transform: uppercase;
}

h1.win {
  color: #5a8a5a;
}

h1.tie {
  color: #000;
}

h1.lose {
  color: #D54349;
}

button {
  background-color: #717171;
  border-radius: 2px;
  border: none;
  color: #eaeaea;
  -webkit-box-flex: 1;
          -ms-flex: 1 1 100%;
              flex: 1 1 100%;
  font-size: 14px;
  margin-right: 2px;
  padding: 18px 18px;
  text-transform: capitalize;
}

button::before {
  border-radius: 20px;
  border: 30px solid #eaeaea;
  content: '';
  display: block;
  margin-bottom: 5px;
}

button.rock::before {
  border-radius: 20px;
}

button.paper::before {
  border-radius: 2px;
}

button.scissors::before {
  border-top: 30px solid rgba(255, 225, 225, 0);
  border-bottom-left-radius: 20px;
  border-bottom-right-radius: 20px;
}

button:focus {
  outline: none;
}

button:active {
  background-color: #5a8a5a;
}

button:hover {
  background-color: #000;
  -webkit-transition: all 0.3s;
  -o-transition: all 0.3s;
  transition: all 0.3s;
}

.app {
  margin: 50vh auto 0;
  -webkit-transform: translateY(-60%);
      -ms-transform: translateY(-60%);
          transform: translateY(-60%);
  width: 300px;
}

.buttons {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  margin: 0 auto;
}

.outcome-display {
  display: block;
}