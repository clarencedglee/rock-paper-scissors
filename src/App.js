//@flow
import 'normalize.css';
import './App.css';

import React, { Component, Fragment } from 'react';
import { options, randomOption, xBeatsY } from './rockPaperScissorsLogic';
import type { Option } from './rockPaperScissorsLogic';
import OutcomeDisplay from './OutcomeDisplay';

type Props = {};
type State = {
  user: Option,
  computer: Option,
  outcome: -1 | 0 | 1 | typeof undefined
};

class App extends Component<Props, State> {
  state = {
    user: 'rock',
    computer: 'scissors',
    outcome: undefined,
  };

  render() {
    const {user, computer, outcome} = this.state
    const outcomeClass = ['lose', 'tie', 'win'][outcome + 1]

    return (
      <div className="app">
        <h1 className={outcomeClass}>
          {user}
          <OutcomeDisplay outcome={outcome}>
            <Fragment>paper</Fragment>
            <Fragment>loses to</Fragment>
            <Fragment>ties with</Fragment>
            <Fragment>beats</Fragment>
          </OutcomeDisplay>
          {computer}
        </h1>

        <OutcomeDisplay outcome={outcome}>
          <Fragment>Pick Rock, Paper, or Scissors below.</Fragment>
          <Fragment>Computer wins...</Fragment>
          <Fragment>Tie</Fragment>
          <Fragment>You win!</Fragment>
        </OutcomeDisplay>

        <div className="buttons">
          {options.map((option, i) =>
            <button onClick={this.choose(option)} className={option} key={i}>
              {option}
            </button>
          )}
        </div>
      </div>
    )
  };

  /**
   * Handle a choice from the user
   * Keep the event element general for now,
   * since we don't use it in our logic
   */
  choose = (option: Option) => 
           (event:  SyntheticEvent<HTMLElement>) =>
  {
    const user     = option
    const computer = randomOption()
    const outcome  = xBeatsY(user, computer)
    this.setState({user, computer, outcome})
  };
};

export default App
