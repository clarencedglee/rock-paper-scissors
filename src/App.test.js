import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import renderer from 'react-test-renderer';

jest.mock('./rockPaperScissorsLogic', () => ({
  // always...
  // lose if x is rock, 
  // tie  if x is paper
  // win  if x is scissors
  xBeatsY: (x, y, options) => ({rock: -1, paper: 0, scissors: 1}[x]),
  randomOption: () => 'rock',
  options: ['rock', 'paper', 'scissors']
}))

it('renders initial state', () => {
  const component = renderer.create(<App/>);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders lose state', () => {
  const component = renderer.create(<App/>);
  component.getInstance().choose('rock')()
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders tie state', () => {
  const component = renderer.create(<App/>);
  component.getInstance().choose('paper')()
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders win state', () => {
  const component = renderer.create(<App/>);
  component.getInstance().choose('scissors')()
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});