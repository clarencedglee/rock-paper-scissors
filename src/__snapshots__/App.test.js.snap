// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders initial state 1`] = `
<div
  className="app"
>
  <h1
    className={undefined}
  >
    rock
    <span
      className="outcome-display"
    >
      paper
    </span>
    scissors
  </h1>
  <span
    className="outcome-display"
  >
    Pick Rock, Paper, or Scissors below.
  </span>
  <div
    className="buttons"
  >
    <button
      className="rock"
      onClick={[Function]}
    >
      rock
    </button>
    <button
      className="paper"
      onClick={[Function]}
    >
      paper
    </button>
    <button
      className="scissors"
      onClick={[Function]}
    >
      scissors
    </button>
  </div>
</div>
`;

exports[`renders lose state 1`] = `
<div
  className="app"
>
  <h1
    className="lose"
  >
    rock
    <span
      className="outcome-display"
    >
      loses to
    </span>
    rock
  </h1>
  <span
    className="outcome-display"
  >
    Computer wins...
  </span>
  <div
    className="buttons"
  >
    <button
      className="rock"
      onClick={[Function]}
    >
      rock
    </button>
    <button
      className="paper"
      onClick={[Function]}
    >
      paper
    </button>
    <button
      className="scissors"
      onClick={[Function]}
    >
      scissors
    </button>
  </div>
</div>
`;

exports[`renders tie state 1`] = `
<div
  className="app"
>
  <h1
    className="tie"
  >
    paper
    <span
      className="outcome-display"
    >
      ties with
    </span>
    rock
  </h1>
  <span
    className="outcome-display"
  >
    Tie
  </span>
  <div
    className="buttons"
  >
    <button
      className="rock"
      onClick={[Function]}
    >
      rock
    </button>
    <button
      className="paper"
      onClick={[Function]}
    >
      paper
    </button>
    <button
      className="scissors"
      onClick={[Function]}
    >
      scissors
    </button>
  </div>
</div>
`;

exports[`renders win state 1`] = `
<div
  className="app"
>
  <h1
    className="win"
  >
    scissors
    <span
      className="outcome-display"
    >
      beats
    </span>
    rock
  </h1>
  <span
    className="outcome-display"
  >
    You win!
  </span>
  <div
    className="buttons"
  >
    <button
      className="rock"
      onClick={[Function]}
    >
      rock
    </button>
    <button
      className="paper"
      onClick={[Function]}
    >
      paper
    </button>
    <button
      className="scissors"
      onClick={[Function]}
    >
      scissors
    </button>
  </div>
</div>
`;
